### PR TITLE
[Feat] Calculate proof size and number of hashes

### PR DESCRIPTION
**Pull request by noel2004**
This PR is base on #692 and provided required features to calculate the size of proofs and number of hashes

Now it has been implied in the e2e test `test_single_add_instance_e2e` and `fibonacci` bench, obtained following output:

> encoded zkvm proof size: 2756371, hash_num: 4411

The encoding of proof is done by `bincode` to kept the serialization as compact as possible

### CODE REVIEW

File: ceno_zkvm/Cargo.toml [hunk 1]
```diff
@@ -44,6 +44,7 @@ tempfile = "3.14"
 thread_local = "1.1"
 
 [dev-dependencies]
+bincode = "1"
```
Comment (matthiasgoergens) [inline ceno_zkvm/Cargo.toml:47]: We already use serde and rkyv, I don't think we need to introduce yet another serialisation framework?  (Perhaps just use serde?)

File: ceno_zkvm/benches/fibonacci.rs [hunk 2]
```diff
@@ -54,6 +85,10 @@ fn fibonacci_prove(c: &mut Criterion) {
                 format!("fibonacci_max_steps_{}", max_steps),
             ),
             |b| {
+                // use itertools::Itertools;
```
Comment (matthiasgoergens) [inline ceno_zkvm/benches/fibonacci.rs:88]: No need to introduce commented out code, please remove.

File: ceno_zkvm/src/e2e.rs [hunk 3]
```diff
@@ -531,6 +531,7 @@ pub fn run_e2e_verify<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>>(
             .verify_proof_halt(zkvm_proof, transcript, exit_code.is_some())
             .expect("verify proof return with error"),
     );
+
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/e2e.rs:534]: Unrelated change, remove.

File: ceno_zkvm/src/scheme/tests.rs [hunk 4]
```diff
@@ -178,6 +179,12 @@ fn test_rw_lk_expression_combination() {
                 &verifier_challenges,
             )
             .expect("verifier failed");
+
+        drop(v_transcript);
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/scheme/tests.rs:183]: Interesting!  Why do we need the manual drop here?

File: ceno_zkvm/benches/fibonacci.rs [hunk 5]
```diff
@@ -43,6 +44,36 @@ fn setup() -> (Program, Platform) {
 fn fibonacci_prove(c: &mut Criterion) {
     let (program, platform) = setup();
     for max_steps in [1usize << 20, 1usize << 21, 1usize << 22] {
+        // estimate proof size data first
```
Comment (matthiasgoergens) [inline ceno_zkvm/benches/fibonacci.rs:47]: Sorry, I don't understand.  Why do we need this estimate first?

File: ceno_zkvm/benches/fibonacci.rs [hunk 6]
```diff
@@ -43,6 +44,36 @@ fn setup() -> (Program, Platform) {
...
+        let (sanity_check_state, _) = run_e2e_with_checkpoint::<E, Pcs>(
+            program.clone(),
+            platform.clone(),
+            vec![],
+            max_steps,
+            Checkpoint::PrepSanityCheck,
+        );
+        assert!(
```
Comment (matthiasgoergens) [inline ceno_zkvm/benches/fibonacci.rs:55]: You can combine your `assert` and your `if let`, since they are checking for the same condition.

Something like `let (proof, verifier)) = sanity_check_state.expect("PrepSanityCheck does not provide proof and verifier");` should do the trick.  (But I don't care about the specifics.  It's just weird to have the same check twice.)

File: transcript/src/statistics.rs
Comment (matthiasgoergens) [inline transcript/src/statistics.rs:1]: Could you add some comments explaining what we are doing here and why, please?

**Review by matthiasgoergens — COMMENTED** [review]
See https://github.com/scroll-tech/ceno/pull/770 for some groundwork.

File: ceno_zkvm/src/scheme/tests.rs [hunk 7]
```diff
@@ -155,8 +155,9 @@ fn test_rw_lk_expression_combination() {
             .expect("create_proof failed");
 
         // verify proof
+        let static_recorder = StatisticRecorder::new();
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/scheme/tests.rs:158]: What is 'static' about this recorder?

**Review by matthiasgoergens — COMMENTED** [review]
Please see https://github.com/scroll-tech/ceno/pull/771

**Comment by matthiasgoergens** [comment]
I merged in latest master (and resolved the conflict).

File: ceno_zkvm/Cargo.toml (hunk 1 above)
Comment (matthiasgoergens) [reply ceno_zkvm/Cargo.toml:47]: Having said all that, if you want to keep bincode, you can keep it.  The only real downside I see is a slightly longer build time, but it's probably not that bad.

**Review by matthiasgoergens — APPROVED** [review]
Please remove the commented out code etc (see my comments), then we can merge.  Thanks!

File: ceno_zkvm/Cargo.toml (hunk 1 above)
Comment (noel2004) [reply ceno_zkvm/Cargo.toml:47]: If we wish to compare the size of our proof with sp1 directly, bincode would provide a bit more confident to the comparision since sp1 use bincode to serialize their proof.

And yes, we can only apply this crate as a dev-dependency

File: ceno_zkvm/benches/fibonacci.rs (hunk 5 above)
Comment (kunxian-xia) [reply ceno_zkvm/benches/fibonacci.rs:47]: Yeah, I think we should move this to `e2e.rs` after we generated the proof. The benchmark we have below is only used to see if there is progress or regression. The proof size should be identical in multiple proving instances.

File: transcript/src/statistics.rs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+type SharedStatisticRecorder = RefCell<StatisticRecorder>;
+
+impl StatisticRecorder {
+    pub fn new() -> Rc<SharedStatisticRecorder> {
+        Rc::new(RefCell::new(Default::default()))
+    }
+}
+
+#[derive(Clone)]
+pub struct BasicTranscriptWitStat<E: ExtensionField> {
```
[the remark this answers is not in the document]
Comment (matthiasgoergens) [reply transcript/src/statistics.rs]: Most likely!

[the remark this answers is not in the document]
Comment (noel2004) [reply transcript/src/statistics.rs]: Have been fixed in #777
